rclcpp/src/rclcpp/executors/multi_threaded_executor.cpp

```diff
@@ -101,5 +101,8 @@ MultiThreadedExecutor::run(size_t)
         scheduled_timers_.erase(it);
       }
     }
+    // Clear the callback_group to prevent the AnyExecutable destructor from
+    // resetting the callback group `can_be_taken_from`
+    any_exec.callback_group.reset();
```

Comment:
I'm concerned this isn't the right fix, @guillaumeautran why did you not change the destructor of `AnyExecutable` instead?

Just changing the callback group to `nullptr` is fragile. There's nothing in the destructor of `AnyExecutable` which implies the callback group might be reset to affect behavior.

Instead, I think, at the very least, `AnyExecutable` should have had a boolean like `should_reset_can_be_taken_from_automatically` which defaulted to true, and the multi threaded executor could change to false. Then it's clearer from both the multi threaded executor's code and any executable's code what's happening.

Reply:
I was not sure if there was a specific reason why the destructor was setup the way it was and did not want to run the risk of leaving the group "locked". I'll make the requested changes then and push another PR.

Reply:
It's possible that since this exists:

https://github.com/ros2/rclcpp/blob/master/rclcpp/src/rclcpp/executor.cpp#L288

We should just remove the code in the destructor of `AnyExecutable` that changes the `can_be_taken_from` of the callback group.

Reply:
@wjwwood Do you think I can add an argument to the constructor of the `AnyExecutable` class?

Reply:
I would be hesitant from removing the if statement. There is a comment that makes a lot of sense:
https://github.com/ros2/rclcpp/blob/97ed34a042c1d4e1a88ea06986112cbdd05e08a9/rclcpp/src/rclcpp/any_executable.cpp#L34